program/rust/src/tests/test_upd_price_no_fail_on_error.rs

```diff
@@ -0,0 +1,154 @@
+    let funding_account = funding_setup.to_account_info();
+
+    let mut price_setup = AccountSetup::new::<pc_price_t>(&program_id);
+    let mut price_account = price_setup.to_account_info();
+    price_account.is_signer = false;
+    initialize_pyth_account_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+
+    let mut clock_setup = AccountSetup::new_clock();
+    let mut clock_account = clock_setup.to_account_info();
+    clock_account.is_signer = false;
+    clock_account.is_writable = false;
+
+    update_clock_slot(&mut clock_account, 1);
+    populate_instruction(&mut instruction_data, Some(42), Some(9), Some(1));
+
+    // We haven't permissioned the publish account for the price account
+    // yet, so any update should fail silently and have no effect. The
+    // transaction should "succeed".
+    assert!(upd_price_no_fail_on_error(
+        &program_id,
+        &[
+            funding_account.clone(),
+            price_account.clone(),
+            clock_account.clone()
+        ],
+        &instruction_data
+    )
+    .is_ok());
+
+
+    {
+        let mut price_data = load_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+        assert_eq!(price_data.comp_[0].latest_.price_, 0);
+        assert_eq!(price_data.comp_[0].latest_.conf_, 0);
+        assert_eq!(price_data.comp_[0].latest_.pub_slot_, 0);
+        assert_eq!(price_data.comp_[0].latest_.status_, PC_STATUS_UNKNOWN);
+        assert_eq!(price_data.valid_slot_, 0);
+        assert_eq!(price_data.agg_.pub_slot_, 0);
+        assert_eq!(price_data.agg_.price_, 0);
+        assert_eq!(price_data.agg_.status_, PC_STATUS_UNKNOWN);
+
+        // Now permission the publish account for the price account.
+        price_data.num_ = 1;
+        pubkey_assign(
+            &mut price_data.comp_[0].pub_,
+            &funding_account.key.to_bytes(),
+        );
+    }
+
+    // The update should now succeed, and have an effect.
+    assert!(upd_price_no_fail_on_error(
+        &program_id,
+        &[
+            funding_account.clone(),
+            price_account.clone(),
+            clock_account.clone()
+        ],
+        &instruction_data
+    )
+    .is_ok());
+
+    {
+        let price_data = load_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+        assert_eq!(price_data.comp_[0].latest_.price_, 42);
+        assert_eq!(price_data.comp_[0].latest_.conf_, 9);
+        assert_eq!(price_data.comp_[0].latest_.pub_slot_, 1);
+        assert_eq!(price_data.comp_[0].latest_.status_, PC_STATUS_UNKNOWN);
+        assert_eq!(price_data.valid_slot_, 0);
+        assert_eq!(price_data.agg_.pub_slot_, 1);
+        assert_eq!(price_data.agg_.price_, 0);
+        assert_eq!(price_data.agg_.status_, PC_STATUS_UNKNOWN);
+    }
+
+    // Invalid updates, such as publishing an update for the current slot,
+    // should still fail silently and have no effect.
+    populate_instruction(&mut instruction_data, Some(55), Some(22), None);
```

Comment:
(this is another place where i think it would be smart to include an assert to validate that someone hasn't accidentally broken the test via a code change. I.e., check that calling `update_price` returns an error but `update_price_no_fail_on_error` doesn't.)

Reply:
Yes!